File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/security/MaintainUserCheckTest.java
package uk.gov.justice.digital.hmpps.oauth2server.security;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.Group;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository;

import java.util.Optional;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThatCode;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class MaintainUserCheckTest {

    @Mock
    private UserRepository userRepository;

    private MaintainUserCheck maintainUserCheck;

    private static final Set<GrantedAuthority> SUPER_USER = Set.of(new SimpleGrantedAuthority("ROLE_MAINTAIN_OAUTH_USERS"));
    private static final Set<GrantedAuthority> GROUP_MANAGER = Set.of(new SimpleGrantedAuthority("ROLE_AUTH_GROUP_MANAGER"));


    @Before
    public void setUp() {
        maintainUserCheck = new MaintainUserCheck(userRepository);
    }

    @Test
    public void superUserDoesNotThrowError()  {
        final var user = User.of("user");
        assertThatCode(() -> maintainUserCheck.ensureUserLoggedInUserRelationship("SuperUser",SUPER_USER,user)).doesNotThrowAnyException();
    }

    @Test
    public void groupManagerGroupInCommonWithUserDoesNotThrowError()  {
        final var user = User.of("user");
        final var group1 = new Group("group", "desc");
        user.setGroups(Set.of(group1, new Group("group2", "desc")));
        final var groupManager = User.of("groupManager");
        groupManager.setGroups(Set.of(new Group("group3", "desc"), group1));
        when(userRepository.findByUsernameAndMasterIsTrue(anyString()))
                .thenReturn(Optional.of(groupManager));

        assertThatCode(() -> maintainUserCheck.ensureUserLoggedInUserRelationship("GroupManager",GROUP_MANAGER,user)).doesNotThrowAnyException();

        verify(userRepository, times(1)).findByUsernameAndMasterIsTrue(anyString());
    }

    @Test
    public void groupManagerNoGroupInCommonWithUserThrowsError()  {
        final var user = User.of("user");
        final var optionalUserEmail = createUser();
        when(userRepository.findByUsernameAndMasterIsTrue(anyString())).thenReturn(optionalUserEmail);

        assertThatThrownBy(() -> maintainUserCheck.ensureUserLoggedInUserRelationship("GroupManager",GROUP_MANAGER,user)).
                isInstanceOf(MaintainUserCheck.AuthUserGroupRelationshipException.class).hasMessage("Unable to maintain user: user with reason: User not with your groups");

        verify(userRepository, times(1)).findByUsernameAndMasterIsTrue(anyString());
    }

    private Optional<User> createUser() {
        return Optional.of(User.of("someUser"));
    }

}

File: src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/MaintainUserCheckTest.kt
package uk.gov.justice.digital.hmpps.oauth2server.security

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThatCode
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.verify
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.Group
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserHelper.Companion.createSampleUser
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository
import uk.gov.justice.digital.hmpps.oauth2server.security.MaintainUserCheck.AuthGroupRelationshipException
import uk.gov.justice.digital.hmpps.oauth2server.security.MaintainUserCheck.AuthUserGroupRelationshipException
import java.util.Optional

class MaintainUserCheckTest {
  private val userRepository: UserRepository = mock()
  private val maintainUserCheck = MaintainUserCheck(userRepository)

  @Test
  fun `superUser Does Not Throw Error when ensureUserLoggedInUserRelationship`() {
    val user = createSampleUser("user")
    assertThatCode {
      maintainUserCheck.ensureUserLoggedInUserRelationship(
        "SuperUser",
        SUPER_USER,
        user
      )
    }.doesNotThrowAnyException()
  }

  @Test
  fun groupManagerGroupInCommonWithUserDoesNotThrowError() {
    val group1 = Group("group", "desc")
    val user = createSampleUser(username = "user", groups = setOf(group1, Group("group2", "desc")))
    val groupManager = createSampleUser("groupManager", groups = setOf(Group("group3", "desc"), group1))
    whenever(userRepository.findByUsernameAndMasterIsTrue(anyString()))
      .thenReturn(Optional.of(groupManager))
    assertThatCode {
      maintainUserCheck.ensureUserLoggedInUserRelationship(
        "GroupManager",
        GROUP_MANAGER,
        user
      )
    }.doesNotThrowAnyException()
    verify(userRepository).findByUsernameAndMasterIsTrue(anyString())
  }

  @Test
  fun groupManagerNoGroupInCommonWithUserThrowsError() {
    val user = createSampleUser("user")
    val optionalUserEmail = createUser()
    whenever(userRepository.findByUsernameAndMasterIsTrue(anyString())).thenReturn(optionalUserEmail)
    assertThatThrownBy {
      maintainUserCheck.ensureUserLoggedInUserRelationship(
        "GroupManager",
        GROUP_MANAGER,
        user
      )
    }.isInstanceOf(AuthUserGroupRelationshipException::class.java)
      .hasMessage("Unable to maintain user: first last with reason: User not with your groups")
    verify(userRepository).findByUsernameAndMasterIsTrue(anyString())
  }

  @Test
  fun `superUser Does Not Throw Error when ensureMaintainerGroupRelationship`() {
    assertThatCode {
      maintainUserCheck.ensureMaintainerGroupRelationship(
        "SuperUser",
        "group1",
        SUPER_USER,
      )
    }.doesNotThrowAnyException()
  }

  @Test
  fun `Group manager able to maintain group`() {
    val groupManager = createSampleUser("groupManager", groups = setOf(Group("group1", "desc"), Group("group2", "desc")))
    whenever(userRepository.findByUsernameAndMasterIsTrue(anyString()))
      .thenReturn(Optional.of(groupManager))
    assertThatCode {
      maintainUserCheck.ensureMaintainerGroupRelationship(
        "GroupManager",
        "group1",
        GROUP_MANAGER,
      )
    }.doesNotThrowAnyException()
    verify(userRepository).findByUsernameAndMasterIsTrue(anyString())
  }

  @Test
  fun `Group manager does not have group so cannot maintain`() {
    val groupManager = createSampleUser("groupManager", groups = setOf(Group("group1", "desc"), Group("group2", "desc")))
    whenever(userRepository.findByUsernameAndMasterIsTrue(anyString()))
      .thenReturn(Optional.of(groupManager))
    assertThatThrownBy {
      maintainUserCheck.ensureMaintainerGroupRelationship(
        "GroupManager",
        "group3",
        GROUP_MANAGER,
      )
    }.isInstanceOf(AuthGroupRelationshipException::class.java)
      .hasMessage("Unable to maintain group: group3 with reason: Group not with your groups")
    verify(userRepository).findByUsernameAndMasterIsTrue(anyString())
  }

  private fun createUser() = Optional.of(createSampleUser("someuser"))

  companion object {
    private val SUPER_USER: Set<GrantedAuthority> = setOf(SimpleGrantedAuthority("ROLE_MAINTAIN_OAUTH_USERS"))
    private val GROUP_MANAGER: Set<GrantedAuthority> = setOf(SimpleGrantedAuthority("ROLE_AUTH_GROUP_MANAGER"))
  }
}
